### ISSUE TITLE
UUID v1 still reports frozen time after exiting freezegun context

### ISSUE THREAD

**Issue by awbacker** [spulec/freezegun#196]
I ran across this issue with `uuid1` because I actually verify the timestamp in the uuids during certain operations and generate a new one if its not in the expected range.   
```
def test():
    import uuid, random, datetime
    from freezegun import freeze_time
    def _show_time():
        print(datetime.datetime.fromtimestamp((uuid.uuid1().time - 0x01b21dd213814000) * 100 / 1e9))
    _show_time()
    frozenat = "2088-08-%02d %02d:11:13" % (random.randint(1, 28), random.randint(1, 23))
    print("freezing at %s" % frozenat)
    with freeze_time(frozenat):
        _show_time()
    _show_time()
```

Paste the above function into a REPL 
```
$> python
Python 2.7.13 (default, May  9 2017, 16:47:31)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> {{paste function}}
>>> test()
2017-07-05 15:26:58.261585
freezing at 2088-08-21 12:11:13
2088-08-21 13:11:13
2088-08-21 13:11:13
>>> test()
2088-08-21 13:11:13
freezing at 2088-08-13 19:11:13
2088-08-21 13:11:13
2088-08-21 13:11:13
```

I'm not sure about that 1 hour difference there (12 vs 13), but its consistent.  I assume its the math above being wrong somehow for py27.  In any case, I validate +- one year so its not important.

I saw this during my unit tests.  I wraped one statement (an api `post`) with freezegun and the all subsequent tests would fail due to the timestamp validation and uuid1() picking up the value used in the previous test.

**Comment by spulec** [comment]
Thanks for opening.

One interesting thing: this works if you change `frozenat` to something like `2017-01-01 01:23:45`. I've dug a bit, but I'm not sure why changing that would break the unpatching of uuid.

**Comment by awbacker** [comment]
This is the output for 4 or 5 different time values I tried using the script [fg.py](https://github.com/spulec/freezegun/files/1127787/fg.txt)
 like `python fg.py "2088-01-01 01:23:45"`

```
 2088-01-01 01:23:45
 ---------------------------------
 before  : 2017-07-06 14:15:26.171666
 during  : 2088-01-01 02:23:45
 after   : 2088-01-01 02:23:45

 2066-01-01 01:23:45
 ---------------------------------
 before  : 2017-07-06 14:15:26.362340
 during  : 2066-01-01 02:23:45
 after   : 2066-01-01 02:23:45

 2018-01-01 01:23:45
 ---------------------------------
 before  : 2017-07-06 14:15:26.510006
 during  : 2018-01-01 02:23:45
 after   : 2018-01-01 02:23:45

 2017-01-01 01:23:45
 ---------------------------------
 before  : 2017-07-06 14:15:26.652049
 during  : 2017-07-06 14:15:26.652049
 after   : 2017-07-06 14:15:26.671410

 2000-01-01 01:23:45
 ---------------------------------
 before  : 2017-07-06 14:15:26.792602
 during  : 2017-07-06 14:15:26.792602
 after   : 2017-07-06 14:15:26.808358
```

**Comment by awbacker** [comment]
uuid1() is trying to prevent duplicating the guid, but its not taking into account bad system clocks or our tricks.
`uuid1()` :
```
current = gen_posix_ts
if current <= last_timestamp:
    current = last_timestamp +1
last_timestamp = current
return new_uuid_using(current)
```
So, there is really no way around this without patching up the uuid1 call (or that module in general) since it is keeping track of things itself.  I think the simplest would simply be to set uuid._last_timestamp = None on start/stop of freezegun so it doesn't play stupid games like this.

I am pretty sure that a misbehaving system clock could also land us in hot water here, at least theoretically, so this doesn't look like a real problem with freezegun itself.

**Comment by awbacker** [comment]
Any chance of getting this (and others?) merged and deploying a 0.4? I can't use it in the current state.

**Comment by spulec** [comment]
Closed with https://github.com/spulec/freezegun/pull/197
